gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java

```diff
@@ -0,0 +1,126 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.InternalApi;
+import com.google.api.gax.rpc.mtls.MtlsProvider;
+import com.google.auto.value.AutoValue;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import javax.annotation.Nullable;
+
+/** Contains the fields required to resolve the endpoint */
+@InternalApi
+@AutoValue
+public abstract class EndpointContext {
+  @Nullable
+  public abstract String clientSettingsEndpoint();
+
+  @Nullable
+  public abstract String transportChannelEndpoint();
+
+  @Nullable
+  public abstract String mtlsEndpoint();
+
+  public abstract boolean switchToMtlsEndpointAllowed();
+
+  @Nullable
+  public abstract MtlsProvider mtlsProvider();
+
+  public abstract Builder toBuilder();
+
+  private String resolvedEndpoint;
+
+  public static Builder newBuilder() {
+    return new AutoValue_EndpointContext.Builder().setSwitchToMtlsEndpointAllowed(false);
+  }
+
+  @VisibleForTesting
+  void determineEndpoint() throws IOException {
+    MtlsProvider mtlsProvider = mtlsProvider() == null ? new MtlsProvider() : mtlsProvider();
+    String customEndpoint =
+        transportChannelEndpoint() != null ? transportChannelEndpoint() : clientSettingsEndpoint();
+    resolvedEndpoint =
+        mtlsEndpointResolver(
+            customEndpoint, mtlsEndpoint(), switchToMtlsEndpointAllowed(), mtlsProvider);
+  }
+
+  // This takes in parameters because determineEndpoint()'s logic will be updated
+  // to pass custom values in.
+  // Follows https://google.aip.dev/auth/4114 for resolving the endpoint
+  @VisibleForTesting
+  String mtlsEndpointResolver(
+      String endpoint,
+      String mtlsEndpoint,
+      boolean switchToMtlsEndpointAllowed,
+      MtlsProvider mtlsProvider)
+      throws IOException {
+    if (switchToMtlsEndpointAllowed && mtlsProvider != null) {
+      switch (mtlsProvider.getMtlsEndpointUsagePolicy()) {
+        case ALWAYS:
+          return mtlsEndpoint;
+        case NEVER:
+          return endpoint;
+        default:
+          if (mtlsProvider.useMtlsClientCertificate() && mtlsProvider.getKeyStore() != null) {
+            return mtlsEndpoint;
+          }
+          return endpoint;
+      }
+    }
+    return endpoint;
+  }
+
+  public String resolveEndpoint() throws IOException {
```

Comment:
Can we try to resolve the endpoint in the `build()` method and just expose a getter for the resolved endpoint?

Reply:
I'm using Autovalue's Builder which I believe generates the `build()` implementation for us. I could create my own Builder class if that's preferable.

Reply:
Yeah, it's a common practice to validate or normalize a field value in the `build()` method, see internal docs [here](go/autovalue/builders-howto#normalize). We have [examples](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPoolSettings.java#L163-L181) in our repo as well.